### ISSUE TITLE
Unable to to connect to wamp router through an nginx proxy

### ISSUE THREAD

**Issue by petten** [KSDaemon/wampy.js#76]
A command-line tool (included in `wampcc`) was able to make a connection through `nginx` to a router (`bonefish`).

The router is on `ws://...` only, but the `nginx` is able to convert `wss://...` to `ws://...` and then back in the reverse direction.

Anyone out there tried something similar?

**Comment by petten** [comment]
The relevant `nginx` configuration looks like this:

```
http {
    server {
        listen 8888;    # client uses this wss_port, instead of 9999 (bonefish one) below
        server_name 10.0.178.178;

        ssl on;
        ssl_certificate <full path to cert file>;
        ssl_certificate_key <full path to cert_key file>;

        location / {
            proxy_pass http://10.0.178.178:9999;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
    }
}
```

When making a direct connection to the router, one uses `ws://10.0.178.178:9999/`.

When making a proxied connection to the router, one uses `wss://10.0.178.178:8888/`.

Again, the `admin` tool found in `wampcc` works going through the `nginx` proxy, whereas `wampy.js` in my tries does not work.

**Comment by KSDaemon** [comment]
Hi! Well, i have similar configuration, and everything works as expected. I’m afk, so l’ll try to take a look asap.

**Comment by KSDaemon** [comment]
@petten Di you try to connect from browser or node.js env?

**Comment by KSDaemon** [comment]
In node environment there was a problem with self signed certificates. I've fixed that. Now you can provide any options for tls.connect(), like ignoring untrusted certs, providing certificates for comparing and etc.

**Comment by petten** [comment]
Thank you Konstantin, for having answered my question so quickly.

I checked my environment, and I did not see any reference to `node.js`.

I was using `wamp.js` from a browser (specifically, `firefox` on `ubuntu`), like this:

```
  ....
  var wampyOpts = {
    realm: "myRealm",
    onConnect: function() { console.log('yahoo! we are online!'); },
    onClose: function() { console.log('see you next time!'); },
    onError: function() { console.log('breakdown occurred'); },
    onReconnect: function() { console.log('reconnecting...'); },
    onReconnectSuccess: function() { console.log('reconnected!'); }
  };

  wampySess = new Wampy("wss://10.0.178.178:8888", wampyOpts);
  ....
```
The callback I got was the `onError`.

Prior to this `nginx` experiement, I had been happily using `wamp.js` for quite a while now, without going through the `nginx`, but directly to a `bonefish` router, like this:

```
  ....
  //| the same wampOpts as the one above

  wampySess = new Wampy("ws://10.0.178.178:9999", wampyOpts);
  ....
```

And I have been getting the `onConnect` callback virtually all the time, unless something is messed up in my own `js` code.

I shall give your fix a try, and report the result back here.

**Comment by petten** [comment]
Additional info about my setup:

- wamp caller: `js` code running in browser using `wampy.js`, in front of an `nginx` proxy (`wss` port `8888`)
- wamp callee: `C++` app using the `wampcc` library, also in front of the `nginx` proxy
- wamp router: `bonefish` (`ws` port `9999`), behind the `nginx` proxy

For the `wampcc` callee, it's possible to use either its `tcp_socket` (directly connecting to the router port `9999`) or `ssl_socket` (connecting to the router through the proxy port `8888`), with successful registration of an RPC procedure at the router.

No matter whether the callee is in front of or behind the `nginx`, I was able to use the command-line util/tool found in the `wampcc` library as a caller to make calls through either the router directly or indirectly through the `nginx`, and get an expected response from the callee.

In summary, all the four scenarios below are successful:

```
 1) caller (util based on wampcc) <-> router (bonefish, 9999) <-> callee (app based on wampcc)

 2) callee <-> nginx (8888) <-> router <-> caller

 3) caller <-> nginx <-> router <-> callee

 4) caller <->
               nginx <-> router
    callee <->
```

Note that the last scenario 4) is the planned final setup, i.e., both the caller and callee are in front of the `nginx` with the router behind it. Scenario 1) has long been used in development and testing, of course.

`wamp.js` as caller so far works in scenarios 1) and 2), but not 3) and 4).

**Comment by KSDaemon** [comment]
@petten A few more questions, regarding your setup:
1. html page is served from different uri, rather than wamp router? Is it on same domain/port/schema?
2. How wamp connection is initialized? I mean: is it done in script, embedded into page, or after some user interactions? 
3. Do you use self-signed certificate?

For test purpose, try to load page via https on same domain/port (it's important) but may be with different path (thats not matter), add button to page, and try to connect to wamp only after pressing button. In this case browser will alert you about self-signed certificate, which you accept, and after that connection to wamp router via wss on same domain(host) and port will go smoothly.

**Comment by petten** [comment]
1. Yes, the html page is served on a machine different from the one running the router. The `nginx` proxy runs on the same machine as the router. In my case, the html page resides on disk, so it's tested as `file:///...` in the browser. I think you are probably right that perhaps I should put it in a place where I can access it via `https://...`.
1. The wamp connection is initiated by a `new Wampy(...);` statement after clicking on a button on the html page. All the js code (including `wampy.js`) is embedded on the page.
1. Yes, I believe the certificate was self-signed.

Perhaps it's possible to change the browser's security configuration settings so that it does not alert the user (popping up a dialog and asking the user to accept...)?

You mentioned in one of your comments yesterday about an option that would make "ignoring untrusted certs" possible. I looked at the documentation about options that may be passed to `new Wamp(...)` or via the `ws.opions(...)` member call, but could not find one that would do it.

**Comment by petten** [comment]
BTW, it's necessary for the `nginx` proxy and wamp router to be running on a separate server, because they are shared by many clients/users out there using their own browsers on their own machines with an http server running.

There are a few things I can try, including getting my html page to run from an `https://...` location.

**Comment by KSDaemon** [comment]
Okay! Now its clear. The problem in your case is that you are trying to connect to WSS with untrusted certificate on different host:port. But browser can not alert you about insecure certificate during wss connection, like it do for normal https page serving. 

Regarding possible options for untrusted connection — it is available only for node.js environment :( Websocket object in browsers does not have such option.

Try, what i say: serve html page and ws proxy url from same domain:port. This allows you to accept an insecure certificate during page loading, and than wss connection (with just different path on same server) will go smoothly.

**Comment by petten** [comment]
I just tried having the html page served by an http server (different from the host where the browser is used), getting the same error (`Firefox can't establish ...`). (The `nginx` proxy and wamp router are still running on a different, 3rd host.)

Next I'll try running the `nginx` proxy on the same host as the http server, leaving the wamp router still on the 3rd host.

**Comment by petten** [comment]
The same error (`Firefox can't establish ...`) occurred.

The html page is served at `http://<the_host>/...` (not sure what kind of http server it is, probably `apache`).

The websocket proxy is `nginx` at `wss://<the_host>:8888` with configuration for it to proxy the wamp router on the 3rd host (`ws://<3rd_host>:9999`), similar to when the two were on the same 3rd host.

**Comment by petten** [comment]
The next thing I'll try is to have the html page served at `https://<the_host>/...` (using default `https` port `443`). Previously it was at the default `http` port `80`.

Your suggestion was to "serve html page and ws proxy url from same domain:port." Does the port really have to be the same for both?

**Comment by petten** [comment]
OK, I got it to work at my original `file:///...` location by first moving from `Firefox` to `Chrome` (`v66`) and then downgrading it to `v48`. (`Firefox` is in general tighter than older versions of `Chrome` in terms of treatment of certificates.)

The issue appears to have always been the certificate that the (newer versions of) browsers do not like one way or another. There was even an command-line option of `--ignore-certificate-errors` recognized by older versions of `Chrome`, but no longer by new ones, not even `v48`. But at least for the particular certificate file that I use, `v48` let it pass; though neither does `v66` nor `Firefox`.

To make it work ultimately using the newer browsers, I guess one must get the certificate generated right, or come up with a better workaround.

**Comment by petten** [comment]
Observed a difference between direct connection to the wamp router and indirect one through the proxy. Periodically, `wampy.js` makes attempts to reconnect to the router in the latter case. Successfully every time so far.

This may not necessarily be attributed to `wampy.js`. Rather, it may very well be the browser's fault.

In comparison, in both cases (direct or indirect), the callee based on the `wampcc` library does not need to reconnect periodically, which I think is the right behavior.
